Iron

xsebek opened the pull request:
- add iron ore, iron mine and iron vein (closes #93)
  - split gear into iron/wooden gear
  - add metal drill
  - add faster recipes with the metal drill
- add compass (closes #341)
- handle multiple entities providing the same capability
  - try to find if the robot has at least one entity providing the capability
  - when no entity could provide the capability rejects it too
- list required devices in the `Incapable` error (closes #342)

byorgey commented on data/entities.yaml:
```diff
@@ -137,6 +137,50 @@
+    char: '■'
+  description:
+  - Raw iron ore. Used to create more resilient tools.
+  - It can only be mined by drilling in the mountains.
```
Seems like this description got copied from `iron ore` and never edited?

xsebek replied:
Fixed :+1:

byorgey commented on data/entities.yaml:
```diff
@@ -137,6 +137,50 @@
+
+- name: iron ore
+  display:
+    attr: copper
```
```suggestion
    attr: iron
```

byorgey commented on data/entities.yaml:
```diff
@@ -137,6 +137,50 @@
+- name: iron ore
+  display:
+    attr: copper
+    char: 'C'
```
```suggestion
    char: 'F'
```

byorgey commented on data/entities.yaml:
```diff
@@ -268,14 +323,23 @@
 
+- name: iron gear
+  display:
+    attr: wood
```
```suggestion
    attr: iron
```

byorgey commented on data/entities.yaml:
```diff
@@ -137,6 +137,50 @@
+    first drilled through, so that the vein becomes accessible.
+  properties: [unwalkable]
+
+- name: iron mine
```
So one idea I had was to get rid of the distinction between `copper vein`/`iron vein` and `copper mine`/`iron mine` and just have `vein` and `mine`.  Given #372, the idea is that drilling in a mine would give you copper with some probability and iron with another (smaller) probability. We can also easily do things like a wooden drill might get you copper but never iron, whereas an iron drill could get you both, etc. --- we just have to specify the right recipes. While this scenario of getting everything from a single mine may not be very accurate in the "real world", I like it for a few reasons:

- It keeps things simpler in terms of the number of different entities in the world.
- It leaves the door open for adding other mine-able things (gold? diamonds? etc.) with much lower probabilities, without having to add yet more separate types of veins and mines.
- With multiple different types of mines, the difficulty for a player lies in locating all the different types and building separate drilling operations at each. With a single mine, the difficulty for the player lies in building a drilling and processing/sorting operation.  The latter sounds much more fun and interesting from a programming perspective than the former.

What do you think?

xsebek replied:
I actually like both ideas - maybe we can split it into different scenarios :thinking: 

It does not require any extra logic in Swarm so the only overhead is extra recipes in our main `data/recipes.yaml`.
I think moving it to some *adventure mode* scenario would both clean up and leave our options open.

My only concern is how to order the changes - I will have to check if scenarios support all I need now or if it is easier to move it over later. :slightly_smiling_face:

xsebek replied:
Alternatively, we could parametrize entities by their underlying material to cut down on boilerplate.

I am not sure how it would work, but maybe it will become more evident once we add gold and other late-game stuff.

byorgey replied:
You said you like both ideas.  What is it you like about the idea of having separate `copper mine`, `iron mine`, etc.?

I don't mind supporting both, and I am not so much concerned about boilerplate as I am about the gameplay experience.  I don't mind having extra entities and recipes in `recipes.yaml` etc.  So we could have both `mine` that works as I described with weighted random results as well as specific `iron mine`, `copper mine`, etc.  The biggest issues then would be (1) which version do you get in the "default" mode?  Even if we support both we still have to decide which is the default game.  And (2) in order to support both we have to allow multiple "world functions" and provide a mechanism for scenarios to specify which world function they will use.  I think this was discussed in an issue somewhere but I can't find it at the moment.

xsebek replied:
I like that it is visual, easier to debug and requires sending robots a bit further. :slightly_smiling_face: 

We could actually do with only mines and have them randomly drilled from `mountain`. That would address the number of entities in the world and make it simpler to add new mines, but it would be even harder to establish mining operations.

byorgey replied:
Hmm, I like the idea of mines appearing from a drilled mountain with a certain probability (in all cases, whether there is only one kind of mine or multiple).  It requires more exploration.  It also feels weird to me that right now the mountains are different colors depending on whether there is a vein --- not that Swarm needs to reflect the real world necessarily, but if you have to drill for something, you should not be able to tell just by looking at the surface what you will find when you drill.

xsebek replied:
I think I found a good balance:
- mountains can be drilled with an 8/1/1 probability to tunnel/copper mine/iron mine
- in scenarios, we can use the deterministic vein ➔ mine custom recipes and entities

byorgey replied:
Ah, OK, I like that.

byorgey commented on src/Swarm/Game/Step.hs:
```diff
@@ -877,8 +878,12 @@ execConst c vs s k = do
+            metalDrill = lookupByName "metal drill" ins
+            insDrill = listToMaybe $ metalDrill <> toyDrill
+
+        drill <- insDrill `isJustOr` Fatal "Drill is required but not installed?!"
```
This should not be a `Fatal` error.  The  `Fatal` error from before was for the situation that looking up `drill` in *the entity map* failed, i.e. the system as a whole did not know about the `drill` entity at all.  But the new code is just checking what kind of drill the robot has installed.  It could certainly happen that the robot is trying to execute a `drill` command even though it has no drill installed.  That is a runtime error but not a `Fatal` one.

xsebek replied:
I suspected as much, but what example would escape our capability checking ahead of drilling?

byorgey replied:
Hmm, good question.  I guess you're right, if this happens it means something has gone wrong with capability checking, so perhaps that should actually be a `Fatal` error.

xsebek replied:
Given that even build inside build checks ahaed, I am going to leave it as Fatal and see if it comes up.

```
> build {r <- build {move; drill down; log ""}; log ""};
<r3> : robot
> salvage
```
```
[busy_williamson] build: You do not have required devices:
  - metal drill or drill
  - logger
```

byorgey reviewed (commented):
Some initial comments.  I haven't finished looking carefully through the changes regarding how we check for required devices etc.

xsebek commented:
Thanks for taking a look @byorgey, the copy-pasting is definitely a problem :sweat_smile: 

I spent most type debugging the capability checking, so now that it (hopefully?) works I will come back and streamline it.

byorgey commented:
By the way, I very much like all the new recipes, the way metal drill does things faster than a regular drill, etc.

xsebek commented:
That Restyled action is a bit annoying since I by mistake pushed to my repo which confused Restyled and it wants to merge to `main`. :facepalm:

byorgey commented on bench/Benchmark.hs:
```diff
@@ -60,13 +60,7 @@ circlerProgram =
-      turn west;
-      move;
-      turn north
+      turn right;
```
Let's change this to `turn right; move; turn right; move; ...` i.e. with the same number of turns and moves as before.  Otherwise the benchmark won't really be comparable to before, since the number of straight-line instructions the robot executes vs the number of recursive calls to `forever` can definitely make a performance difference.

xsebek replied:
Indeed it does seem to be 20% slower, I should have checked :+1:

xsebek replied:
Fixed :heavy_check_mark:

byorgey commented on data/recipes.yaml:
```diff
@@ -220,6 +220,13 @@
+  - [1, water]
+  - [1, box]
+  out:
+  - [1, compass]
```
I like it.

byorgey commented on src/Swarm/Game/Step.hs:
```diff
@@ -374,10 +374,19 @@ stepCESK cesk = case cesk of
+    orient <- hasCapability COrient
+    if isCardinal d && not orient
+      then return $ Up (Incapable (S.singleton COrient) (TDir d)) s []
+      else return $ Out (VDir d) s k
```
I am not sure about this.  Should we really need the `orient` capability just to *evaluate* a cardinal direction constant?

xsebek replied:
Well, `base` would escape the check if I did not do it, but it still seems wrong:
```Haskell
> turn north -- compass is needed (but the error does not mention it)
> let d = north in build {turn d} -- once base has compass, nobody else needs it
```
I probably just overthought it and checking it when turning and drilling would work just fine.

- related #201 for the case when a robot is sent in some direction and it turns out it is cardinal at runtime (I should write a test for it...)
- related error message improvement #342

xsebek replied:
Reverted and moved the check to commands using the direction.

xsebek commented on src/Swarm/Language/Syntax.hs:
```diff
@@ -120,12 +120,12 @@ dirInfo d = case d of
   DLeft -> relative (\(V2 x y) -> V2 (- y) x)
   DRight -> relative (\(V2 x y) -> V2 y (- x))
   DBack -> relative (\(V2 x y) -> V2 (- x) (- y))
+  DDown -> relative (const down)
```
I think I might have at some point advocated for `down` being a cardinal direction, but now I need it to be allowed without a compass :sweat_smile: I do not think it's worth coming up with another name for the set of directions that are cardinal but not `down`. Writing `/= down` everywhere is also an option, but not a good one. :shrug:

byorgey replied:
Yeah, making `down` a relative direction makes sense, given the existence of `compass`.  You should not need a `compass` to find the ground. :laughing:

xsebek commented:
Hey @byorgey, could you please take a look and test if this works?

I made a testing playground with the old `copper vein`, `iron vein` and both drills and `compass` in inventory:
```bash
cabal run swarm:swarm -- --scenario 03Challenges/02-drill_test --cheat
```
![Screenshot from 2022-06-11 02-22-12](https://user-images.githubusercontent.com/44544735/173166780-15fa4a17-65b8-4c00-8f96-9d219d483740.png)

I did not get around to automating the tests yet, but I would be happy to let someone else test my fragile code if they felt like it :wink:

byorgey commented:
The test scenario seems to work.  However, when I started creative mode and tried sending a robot to drill some mountains, it did not work.  Maybe it is failing due to having no `drill` installed even though we are in creative mode?

byorgey commented on src/Swarm/Game/Step.hs:
```diff
@@ -1412,6 +1383,68 @@ execConst c vs s k = do
+        -- device is ok if it is available in the inventory of parent
+        -- when building or installed in target robot when reprogramming
+        deviceOK d = inventory `E.contains` d
+        ignoreOK ([], miss) = ([], miss)
```
Took me a little while to figure out what this is for.  Maybe add a comment, something like: ignoreOK is given two lists of devices which all provide a given capability, a list of OK devices and a list of the other devices. If *at least one* device providing a capability is OK (i.e. the first list is not empty), then we can ignore all the alternative devices that are not installed.

xsebek replied:
Sure, I added some comments :+1:

byorgey commented on src/Swarm/Util/Yaml.hs:
```diff
@@ -61,6 +62,9 @@ liftE = E . const
 withE :: Semigroup e => e -> With e f a -> With e f a
 withE e (E f) = E (f . (<> e))
 
+fromE :: (Monad f) => With e f e
```
I would call this `getE`.  `fromE` sounds like you're converting `e` to some other type.

xsebek replied:
I was considering `askE`, to make it look like `ReaderT`, but `getE` sounds good too.

byorgey replied:
Yeah, `askE` is probably better!

xsebek replied:
Too late, I used your suggestion :laughing:

byorgey replied:
Hah! Either one is fine. 😄

byorgey commented on src/Swarm/Game/Step.hs:
```diff
@@ -1412,6 +1383,68 @@ execConst c vs s k = do
+                    )
+    -- check that there are in fact devices to provide every required capability
+    (creative || not (any null deviceSets))
+      `holdsOrFail` [singularSubjectVerb subject "can", "not perform an impossible task"]
```
To make sure I'm understanding correctly, if this "impossible task" error is triggered, it means that there do not exist any devices to provide the required capability, right?  As in, we have some capability but no actual devices yet in `entities.yaml` which provide it?

byorgey replied:
Hmm, I actually got this error while playing in classic mode and it's very mysterious.  How hard would it be to output the name of the capability for which there is no device, with a message apologizing and a link to the issue tracker?

xsebek replied:
I expected this to only come up for `CGod` tier tasks, like `create "life"`, where the message would make sense.

Maybe I could check if it is `CGod` and otherwise link to the issue. Something like:
```Haskell
[ singularSubjectVerb subject "do", "not have", indefiniteQ capName
, "capability, because no device provides it."
, "\nSee https://github.com/swarm-game/swarm/issues/26."
]
```

However, I feel like it should be `Incapable` exceptions job to figure this out. :smirk:

xsebek replied:
Some of it was common, but building and reprogramming can give a more specialised error message, so I keep some duplicity code. Oh well, at least I get to close #342 :innocent:

byorgey replied:
Also, just an update: the reason I got this error while playing in classic mode was actually due to #389 .  So that was making it more mysterious than necessary.

byorgey reviewed (commented):
Everything looks good to me, but there's still something I don't understand going on with drilling in creative mode.

xsebek commented on data/scenarios/03Challenges/02-drill_test.yaml:
```diff
@@ -0,0 +1,140 @@
+    growth: [0,0]
+    yields: water
+
+  - name: knownwavywater
```
I suggest we solve it later, I made an Issue so we don't forget:
- #387

byorgey replied:
👍

xsebek commented:
> Everything looks good to me, but there's still something I don't understand going on with drilling in creative mode.

@byorgey it was actually more than drilling, by mistake I did not install any required entities if they were not present in inventory.

The commit [9e24293](https://github.com/swarm-game/swarm/pull/373/commits/9e2429394850d9f51a9b857abd46924798512663) should fix that :slightly_smiling_face:

byorgey commented on src/Swarm/Game/Exception.hs:
```diff
@@ -43,18 +75,84 @@ data Exn
+formatIncapable em f caps tm
+  | CGod `S.member` caps =
+    unlinesExText
+      [ "Can not perform an impossible task:"
```
I think this error message could still be improved.  I don't think it is going to make much sense if a player stumbles across it.  *e.g.* imagine that they type `as` accidentally and notice that it has a type, then they wonder what it does so they try it out. They are going to wonder *why* the task is impossible. Somehow we should convey that this is a command that is not available during normal play.  We can definitely make it funny though.  Some ideas:
- "This command is not available to mere mortals" / "This command is forbidden"
- "Thou shalt not utter such blasphemy"
- If we wanted to be more informative and less funny, "This command is only available to system robots and scenario definitions".

xsebek replied:
I guess the Elden ring is now popular, so blasphemy it is:
```Haskell
incapableError (S.singleton CGod) (TConst As)
-- Thee shalt not utter such blasphemy:
--   'as'
--   If't be true thee wanteth to playeth god, then tryeth Creative game.
```

byorgey commented:
This is looking great! :rocket:

byorgey commented:
But up to you whether we want to go ahead and merge this and file a separate issue re the problems I mentioned above!  At this point I would actually lean towards doing it that way rather than hold this up longer.

byorgey commented:
A capability error reporting bug that I am likewise happy to split out into a separate issue after merging this, just want to write it down while it's fresh in my head:

- Start a classic mode game with `--cheat`
- Switch into creative mode and `create "branch predictor"`
- Type `build {move; if (1 > 2) {} {}}`
    - As expected we get an error saying no device provides the `compare` capability.
- Now type `build {move; turn left; if (1 > 2) {} {}}`
    - All we have done is add `turn left`; we would expect to get the same error.  Instead we get an error like `You do not have the devices required for: 'move; turn left; if (1 > 2) {noop} {noop}'  please obtain:   - branch predictor`
    - But we do have a branch predictor, so this error is bogus.  Not sure yet whether the capability checking or error reporting is to blame.

xsebek commented:
@byorgey I think your first bug is present on main too, the Drill capability is not reported by `requiredCaps`.
The difference is, that I call that a `Fatal` error because `build` should have given it to the robot, while `main` drives the robots to rocks and there they stop without a drill.

See the debug logs I added:

### Main

![Screenshot from 2022-06-14 17-01-12](https://user-images.githubusercontent.com/44544735/173621697-64820895-06a5-4480-96c0-ccc56660187c.png)

### Iron
I was also missing standard devices, so I added them back:
![Screenshot from 2022-06-14 16-48-55](https://user-images.githubusercontent.com/44544735/173622060-301ed588-205e-4c6a-bf4f-062086c6b78b.png)
![Screenshot from 2022-06-14 17-19-14](https://user-images.githubusercontent.com/44544735/173622090-0a74e8e9-ef89-408b-9190-30f6e43e5229.png)

---

I have no idea what the second bug is, but I hope it is not really my fault :smile:
